tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/core/SentenceTagger.scala

```diff
@@ -0,0 +1,279 @@
+      case KeywordTaggerInitializer(keywords) =>
+        IndependentTokenSentenceTagger(KeywordTagger(keywords))
+      case GoogleUnigramTaggerInitializer(tagType) =>
+        IndependentTokenSentenceTagger(GoogleUnigramTagger(googleNgrams.get, tagType))
+      case VerbnetTaggerInitializer =>
+        VerbnetTagger(verbnet.get)
+    }
+  }
+
+  /** Applies a sequence of sentence taggers to a sentence, tagging each token with the union
+    * of their tags.
+    *
+    * @param sentence the sentence we want to tag
+    * @param taggers the taggers we want to apply
+    * @return a tagged sentence
+    */
+  def tagWithMultipleTaggers(sentence: Sentence, taggers: Seq[SentenceTagger]): TaggedSentence = {
+    val taggings = taggers map { tagger =>
+      tagger.tag(sentence)
+    }
+    TaggedSentence(
+      sentence,
+      (Range(0, sentence.tokens.size) map { tokIndex =>
+      (tokIndex, taggings.toSet flatMap { tagging: TaggedSentence =>
+        tagging.tags.getOrElse(tokIndex, Set())
+      })
+    }).toMap
+    )
+  }
+
+  /** Tags all sentences from a given sentence source.
+    *
+    * Note that this function is multithreaded.
+    *
+    * @param tagger the tagger to apply
+    * @param sentenceSource the sentence source
+    * @return an iterator over the resulting tagged sentences
+    */
+  def tagSentenceSource(
+    tagger: SentenceTagger,
+    sentenceSource: SentenceSource
+  ): Iterator[TaggedSentence] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    val taggingTasks: Iterator[Future[TaggedSentence]] =
+      for {
+        sentence <- sentenceSource.sentenceIterator
+      } yield Future {
+        tagger.tag(sentence)
+      }
+    val futureTagged: Future[Iterator[TaggedSentence]] = Future.sequence(taggingTasks)
+    Await.result(futureTagged, 2 days)
+  }
+}
+
+/** A SentenceTagger that assumes tokens can be independently tagged.
+  *
+  * @param tokenTagger the token tagger to apply to each token of the sentence
+  */
+case class IndependentTokenSentenceTagger(tokenTagger: TokenTagger) extends SentenceTagger {
+
+  override def tag(sentence: Sentence): TaggedSentence = {
+    TaggedSentence(
+      sentence,
+      (sentence.tokens.zipWithIndex map { _.swap }).toMap mapValues { tok =>
+        tokenTagger.tag(tok)
+      }
+    )
+  }
+}
+
+/** A SentenceTagger that tags tokens at key positions in the sentence (specifically the
+  * first, second, second-last, last, and nexus tokens).
+  */
+case object TokenPositionTagger extends SentenceTagger {
+
+  private val featureName = 'place
+  private val hasNexusSymbol = 'nexus
+  private val hasFirstSymbol = 'first
+  private val hasSecondSymbol = 'second
+  private val hasSecondLastSymbol = 'secondLast
+  private val hasLastSymbol = 'last
+
+  override def tag(sentence: Sentence): TaggedSentence = {
+    TaggedSentence(
+      sentence,
+      (Range(0, sentence.tokens.size) map {
```

Comment:
This is a weird way to do this. Why not just create a mutable indexed seq, and set the appropriate values?